// notmebotz-tools/scrape/search/spotifys.js
 const axios = require('axios');

const CLIENT_ID = '4c4fc8c3496243cbba99b39826e2841f';
const CLIENT_SECRET = 'd598f89aba0946e2b85fb8aefa9ae4c8';

async function getAccessToken() {
    const url = 'https://accounts.spotify.com/api/token';
    const headers = {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Authorization': `Basic ${Buffer.from(`${CLIENT_ID}:${CLIENT_SECRET}`).toString('base64')}`,
    };
    const data = new URLSearchParams({ grant_type: 'client_credentials' });

    const response = await axios.post(url, data, { headers });
    return response.data.access_token;
}

async function searchSong(query) {
    const accessToken = await getAccessToken();
    const url = `https://api.spotify.com/v1/search`;
    const headers = { 'Authorization': `Bearer ${accessToken}` };
    const params = { q: query, type: 'track', limit: 10 };

    const response = await axios.get(url, { headers, params });
    const results = response.data.tracks.items.map(track => ({
        name: track.name,
        artists: track.artists.map(artist => artist.name).join(', '),
        album: track.album.name,
        url: track.external_urls.spotify,
    }));

    return {
        author: "Herza",
        status: 200,
        results
    };
}
module.exports = { searchSong }
