// notmebotz-tools/scrape/misc/catbox.js
const axios = require('axios');
const FormData = require('form-data');
const fs = require('fs');

async function catbox(path) {
    const data = new FormData();
    data.append('reqtype', 'fileupload');
    data.append('userhash', '');
    data.append('fileToUpload', fs.createReadStream(path));

    const config = {
        method: 'POST',
        url: 'https://catbox.moe/user/api.php',
        headers: {
            'User-Agent': 'Mozilla/5.0 (Android 10; Mobile; rv:131.0) Gecko/131.0 Firefox/131.0',
            'Accept': 'application/json',
            'accept-language': 'id-ID',
            'referer': 'https://catbox.moe/',
            'cache-control': 'no-cache',
            'x-requested-with': 'XMLHttpRequest',
            'origin': 'https://catbox.moe',
            'sec-fetch-dest': 'empty',
            'sec-fetch-mode': 'cors',
            'sec-fetch-site': 'same-origin',
            'te': 'trailers',
            ...data.getHeaders()
        },
        data: data
    };
    const api = await axios.request(config);
    return {
        author: "Herza",
        status: 200,
        url: api.data 
    };
}

module.exports = { catbox };
