// notmebotz-tools/scrape/misc/ssweb.js
const { run } = require('shannz-playwright');

async function sswebs(url, type) {
    const code = `const { chromium, devices } = require('playwright');

    const devicesMap = {
        mobile: devices['iPhone 12'],
        tablet: devices['iPad'],
        desktop: null
    };

    async function ssweb(url, type) {
        const device = devicesMap[type];

        const browser = await chromium.launch();
        const context = device ? await browser.newContext({ ...device }) : await browser.newContext();
        const page = await context.newPage();

        await page.goto(url);
        await page.screenshot({ path: 'screenshot.png', fullPage: true });
        console.log('Screenshot saved as screenshot.png');

        await browser.close();
    }
    ssweb(\`${url}\`, \`${type}\`).then(a => console.log(a));`;
    
    const start = await run('javascript', code);
    const result = start.result.files;
    const urls = result[0].publicURL;
    return {
        author: "Herza",
        status: 200,
        data: {
        url: 'https://try.playwright.tech' + urls,
        fileName: result[0].fileName,
        extension: result[0].extension
        }
    }
}

module.exports = { sswebs }
