// notmebotz-tools/scrape/AI/txt2img.js
const axios = require('axios');
const fileType = require('file-type');
const fetch = require('node-fetch');
const FormData = require('form-data');

async function boreal(prompt, negativePrompt = 'low quality') {
    const data = JSON.stringify({
        "model": {
            "instance_prompt": "photo",
            "image": "https://huggingface.co/kudzueye/Boreal/resolve/main/images/ComfyUI_00845_.png",
            "id": "kudzueye/Boreal",
            "metadata": "8cd8fcd4-73b0-4742-b67c-b82625b07b52_metadatakudzueye-Boreal.png",
            "isPublic": true,
            "gallery": [],
            "infos": {}
        },
        "inputs": prompt,
        "parameters": {
            "negative_prompt": negativePrompt
        }
    });

    const config = {
        method: 'POST',
        url: 'https://lorastudio.co/api/generate',
        headers: {
            'User-Agent': 'Mozilla/5.0 (Android 10; Mobile; rv:131.0) Gecko/131.0 Firefox/131.0',
            'Content-Type': 'application/json',
            'accept-language': 'id-ID',
            'referer': 'https://lorastudio.co/generate?model=kudzueye/Boreal',
            'origin': 'https://lorastudio.co',
            'sec-fetch-dest': 'empty',
            'sec-fetch-mode': 'cors',
            'sec-fetch-site': 'same-origin',
            'priority': 'u=0',
            'te': 'trailers'
        },
        data: data
    };

    try {
        const response = await axios.request(config);
        const imageData = response.data.image;

        if (!imageData) {
            throw new Error('Image data not received');
        }

        const base64Data = imageData.replace(/^data:image\/png;base64,/, "");
        const buffer = Buffer.from(base64Data, 'base64');
        return await uploadToCdn(buffer);
    } catch (error) {
        console.error('Error in boreal:', error.message);
        return null;
    }
}

async function uploadToCdn(buffer) {
    try {
        const { ext } = (await fileType.fromBuffer(buffer)) || { ext: 'png' };
        const bodyForm = new FormData();
        bodyForm.append("fileToUpload", buffer, `file.${ext}`);
        bodyForm.append("reqtype", "fileupload");

        const res = await fetch("https://catbox.moe/user/api.php", {
            method: "POST",
            body: bodyForm,
        });

        const data = await res.text();

        return {
            author: "Herza",
            status: 200,
            data
        };
    } catch (error) {
        console.error('Error in uploadToCdn:', error.message);
        return {
            author: "Herza",
            status: 500,
            error: error.message
        };
    }
}

module.exports = { boreal };
