// notmebotz-tools/scrape/stalk/steam.js
/*
*
* [ *SCRAPE STEAM STALK* ]
* Created By Hann
* 
* Channel: https://whatsapp.com/channel/0029Vaf07jKCBtxAsekFFk3i
*
**/
const axios = require('axios');
const cheerio = require('cheerio');

async function steamStalk(teks) {
    try {
        const { data } = await axios.get(`https://steamcommunity.com/id/${teks}`);
        const $ = cheerio.load(data);

        const profileData = {
            personaName: $('.actual_persona_name').text().trim() || "Tidak Memiliki Data",
            realName: $('.header_real_name bdi').text().trim() || "Tidak Memiliki Data",
            location: $('.header_real_name').contents().last().text().trim() || "Tidak Memiliki Data",
            avatar: $('.playerAvatar img').attr('src') || "Tidak Memiliki Data",
            level: $('.friendPlayerLevelNum').text().trim() || "Tidak Memiliki Data",
            badges: {
                total: parseInt($('.profile_count_link_total').first().text().trim(), 10) || "Tidak Memiliki Data",
                badgeInfo: []
            },
            games: parseInt($('.profile_count_link_total').eq(1).text().trim(), 10) || "Tidak Memiliki Data",
            screenshots: parseInt($('.profile_count_link_total').eq(3).text().trim(), 10) || "Tidak Memiliki Data",
            videos: parseInt($('.profile_count_link_total').eq(4).text().trim(), 10) || "Tidak Memiliki Data",
            reviews: parseInt($('.profile_count_link_total').eq(5).text().trim(), 10) || "Tidak Memiliki Data",
            friends: parseInt($('.profile_count_link_total').eq(6).text().trim(), 10) || "Tidak Memiliki Data",
            achievements: {
                total: parseInt($('.showcase_stat').first().text().trim(), 10) || "Tidak Memiliki Data",
                avgCompletionRate: parseInt($('.showcase_stat').eq(1).text().trim().replace('%', '').trim(), 10) || "Tidak Memiliki Data"
            },
            recentGames: [],
            comments: []
        };

        $('.profile_header_badge').each((_, element) => {
            profileData.badges.badgeInfo.push({
                name: $(element).find('.favorite_badge_description .name').text().trim() || "Tidak Memiliki Data",
                xp: $(element).find('.favorite_badge_description .xp').text().trim() || "Tidak Memiliki Data",
                image: $(element).find('.badge_icon').attr('src') || "Tidak Memiliki Data"
            });
        });

        $('.recent_game').each((_, element) => {
            profileData.recentGames.push({
                gameName: $(element).find('.game_name a').text().trim() || "Tidak Memiliki Data",
                lastPlayed: parseFloat($(element).find('.game_info_details').text().trim().split(' ')[0]) || "Tidak Memiliki Data"
            });
        });

        $('.commentthread_comment').each((_, element) => {
            profileData.comments.push({
                author: $(element).find('.commentthread_comment_author a').text().trim() || "Tidak Memiliki Data",
                text: $(element).find('.commentthread_comment_text').text().trim() || "Tidak Memiliki Data",
                timestamp: $(element).find('.commentthread_comment_timestamp').attr('title') || "Tidak Memiliki Data"
            });
        });

        return {
            author: "Herza",
            status: 200,
            data: profileData
        };
    } catch (error) {
        return {
            author: "Herza",
            status: 500,
            data: error.message
        };
    }
}

module.exports = { steamStalk };
