// notmebotz-tools/scrape/aio.js
const axios = require("axios");
const cheerio = require("cheerio");
const FormData = require("form-data");

async function Aio(urls) {
    try {
        const { data: tkns } = await axios.get("https://steptodown.com/");
        const $ = cheerio.load(tkns);
        const token = $("#token").val();

        const formData = new FormData();
        formData.append("url", urls);
        formData.append("token", token);

        const { data: response } = await axios.post(
            "https://steptodown.com/wp-json/aio-dl/video-data/",
            formData,
            {
                headers: {
                    ...formData.getHeaders()
                }
            }
        );

        const processResponse = (data) => {
            if (!data) return {};

            const processValue = (val) => {
                if (Array.isArray(val)) {
                    return val.map(item => processValue(item));
                }
                if (val === null) return null;
                if (typeof val === 'object') {
                    return Object.fromEntries(
                        Object.entries(val).map(([k, v]) => [k, processValue(v)])
                    );
                }
                return val;
            };

            return processValue(data);
        };

        const processedData = processResponse(response);

        return {
            author: "Herza",
            status: 200,
            data: processedData
        }
    } catch (error) {
        return {
            author: "Herza",
            status: 500,
            message: error.message,
            data: null
        };
    }
}

module.exports = { Aio };
