// notmebotz-tools/scrape/fbdl.js
const axios = require('axios');
const cheerio = require('cheerio');
const FormData = require('form-data');

async function fbdl(url) {
  try {
    const form = new FormData();
    form.append('id', url);
    form.append('locale', 'id');

    const response = await axios.post('https://getmyfb.com/process', form, {
      headers: {
        ...form.getHeaders()
      },
    });

    if (response.status === 200) {
      const $ = cheerio.load(response.data);
      const title = $('.results-item-text').text().trim();
      const thumbnail = $('.results-item-image').attr('src') || null;
      const urlHd = $('.results-list li:nth-child(1) a').attr('href') || null;
      const urlSd = $('.results-list li:nth-child(2) a').attr('href') || null;

      return {
        author: "Herza",
        status: 200,
        data: {
        title,
        thumb: thumbnail,
        video: {
          hd: urlHd,
          sd: urlSd,
        }
        }
      };
    } else {
      return {
        author: "Herza",
        status: response.status
      };
    }
  } catch (error) {
    return {
      status: 'error',
      message: error.message
    };
  }
}

module.exports = { fbdl };
