// notmebotz-tools/scrape/ytdl3.js
const ya = require('@justherza/ytdl-me')

const ytdl3 = async (videoUrl, format = "480") => {
  try {
    if (videoUrl.includes("youtu.be")) {
      const videoId = videoUrl.split("youtu.be/")[1];
      videoUrl = `https://www.youtube.com/watch?v=${videoId}`;
    }

    const availableFormats = {
      video: ["480","720","1080","1440","4k","8k"],
      audio: ["mp3","ogg","wav","m4a","webm","aac","opus"],
    };

    const isAudio = availableFormats.audio.includes(format);
    const isVideo = availableFormats.video.includes(format);

    if (!isAudio && !isVideo) {
      return { 
        status: 400, 
        error: `That type is not available, the available formats are: video (${availableFormats.video.join(", ")}) and audio (${availableFormats.audio.join(", ")})` 
      };
    }

    let datanya = await ya.download({
      yt_link: videoUrl,
      yt_format: format,
      logs: false,
      saveId: false
    })

    const result = {
      author: "Herza",
      status: 200,
      data: datanya
    };

    return result;
  } catch (error) {
    console.error("Error:", error.message);
    return { status: 500, error: error.message };
  }
};

module.exports = { ytdl3 }

// Contoh penggunaan:
// let z = await ytdl3("https://www.youtube.com/watch?v=dQw4w9WgXcQ", "720");
// console.log(z);
